lib/fs.js

```diff
@@ -319,39 +329,46 @@ fs.existsSync = function(path) {
   }
 };
 
+const defaultReadFileOpts = Object.create(null, {
+  flag: { value: O_RDONLY, writable: true }
+});
 fs.readFile = function(path, options, callback) {
+  var flag;
+  var encoding;
   callback = maybeCallback(arguments[arguments.length - 1]);
-  options = getOptions(options, { flag: 'r' });
+  if (options != null && typeof options !== 'function') {
+    options = getOptions(options, defaultReadFileOpts);
+    flag = (options.flag ? stringToFlags(options.flag) : O_RDONLY);
+    encoding = options.encoding;
+  } else {
+    flag = O_RDONLY;
+  }
 
-  if (handleError((path = getPathFromURL(path)), callback))
-    return;
+  if (typeof path !== 'string') {
+    if (handleError((path = getPathFromURL(path)), callback))
+      return;
+  }
   if (!nullCheck(path, callback))
     return;
 
-  var context = new ReadFileContext(callback, options.encoding);
-  context.isUserFd = isFd(path); // file descriptor ownership
-  var req = new FSReqWrap();
-  req.context = context;
-  req.oncomplete = readFileAfterOpen;
-
+  var context = new ReadFileContext(callback, encoding, isFd(path));
   if (context.isUserFd) {
-    process.nextTick(function() {
-      req.oncomplete(null, path);
+    process.nextTick(() => {
+      context.req.oncomplete(null, path);
     });
     return;
   }
 
-  binding.open(pathModule._makeLong(path),
-               stringToFlags(options.flag || 'r'),
-               0o666,
-               req);
+  binding.open(pathModule._makeLong(path), flag, 0o666, context.req);
 };
 
-const kReadFileBufferLength = 8 * 1024;
+function ReadFileContext(callback, encoding, isUserFd) {
+  this.req = new FSReqWrap();
```

Comment:
Nice, moving things here looks a lot cleaner. I think any slots saved here could be a big win